### PR TITLE
First attempt at refactoring `Oracle`s

### PR DESCRIPTION
**Pull request by Kukovec**
- [ ] Tests added for any new code
- [x] Ran `make fmt-fix` (or had formatting run automatically on all files edited)
- [ ] Documentation added for any new functionality
- [ ] [Entries added to `./unreleased/`][changelog format] for any new functionality

First step towards #2708. 
The PR makes the following changes to https://github.com/informalsystems/apalache/blob/6bb36acc1ad88fdeba21d45b67663b1e792e29c6/tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/rules/aux/Oracle.scala:

  - `SymbState` is replaced by `RewriterScope`
  - Method names/parameters are renamed to be more verbose, but also more descriptive
  - Exception previously thrown by each implementation individually has been pushed into the base trait
  - Signature of `caseAssertions` has been changed such that the optional sequence is now an `Option`, instead of treating the empty sequence as a special case.
  - Signature of `getIndexOfOracleValueFromModel` (previously `evalPosition`) has been changed, to no longer include `state`. Upon inspection, none of the oracle implementations ever accessed this parameter.
   
`IntOracle` is implemented with the changes above as a prototype. Note that tests are currently missing for the IntOracle, and will be added in a subsequent PR.

[changelog format]: https://github.com/informalsystems/apalache/blob/main/CONTRIBUTING.md#how-to-record-a-change

### CODE REVIEW

File: tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/stratifiedRules/aux/oracles/IntOracle.scala [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+import at.forsyte.apalache.tla.lir.{IntT1, ValEx}
+import at.forsyte.apalache.tla.lir.values.TlaInt
+import at.forsyte.apalache.tla.typecomp.TBuilderInstruction
+import at.forsyte.apalache.tla.types.tla
+
+/**
+ * An oracle that uses an integer variable. Although using integers as an oracle is the most straightforward decision,
+ * do not use this oracle by default. It is handy, when reasoning about sequences.
+ *
+ * TODO: add tests
```
Comment (konnov) [inline tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/stratifiedRules/aux/oracles/IntOracle.scala:15]: How about adding tests? :)

This should be super-easy with the new design. Just a bunch of unit tests

**Review by konnov — COMMENTED** [review]
This is a nice PR. It looks nice, structured, and extremely well commented. How about adding a bunch of unit tests?

File: tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/stratifiedRules/aux/oracles/IntOracle.scala [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+import at.forsyte.apalache.tla.bmcmt.stratifiedRules.RewriterScope
+import at.forsyte.apalache.tla.bmcmt.types.CellT
+import at.forsyte.apalache.tla.lir.{IntT1, ValEx}
+import at.forsyte.apalache.tla.lir.values.TlaInt
+import at.forsyte.apalache.tla.typecomp.TBuilderInstruction
+import at.forsyte.apalache.tla.types.tla
+
+/**
+ * An oracle that uses an integer variable. Although using integers as an oracle is the most straightforward decision,
+ * do not use this oracle by default. It is handy, when reasoning about sequences.
```
Comment (thpani) [inline tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/stratifiedRules/aux/oracles/IntOracle.scala:13]: > Although using integers as an oracle is the most straightforward decision, do not use this oracle by default.

Can we explain why?

File: tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/stratifiedRules/aux/oracles/IntOracle.scala [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+   * Produce an expression that states that the chosen value equals to the value `v_{index}`. The actual implementation
+   * may be different from an integer comparison.
+   */
+  override def oracleValueIsEqualToIndexedValue(scope: RewriterScope, index: Int): TBuilderInstruction =
+    tla.eql(intCell.toBuilder, tla.int(index))
+
+  override def getIndexOfOracleValueFromModel(solverContext: SolverContext): Int =
+    solverContext.evalGroundExpr(intCell.toBuilder) match {
+      case ValEx(TlaInt(i)) => i.toInt
+      case _                => throw new IllegalStateException(s"Invalid call to evalPosition, not an integer.")
```
Comment (thpani) [inline tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/stratifiedRules/aux/oracles/IntOracle.scala:37]: Where does the `evalPosition` stem from?

File: tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/stratifiedRules/aux/oracles/Oracle.scala [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+  def oracleValueIsEqualToIndexedValue(scope: RewriterScope, index: Int): TBuilderInstruction
+
+  /**
+   * Produce a ground expression that contains assertions for the possible oracle values.
+   *
+   * The `assertions` sequence must contain `N` elements, where `N` equals the number of oracle value candidates. The
+   * `i`-th element of `assertions` describes a formula, which holds if the oracle value is equal to the `i`-th
+   * candidate value `vi`.
+   *
+   * Optionally, a sequence option `elseAssertionsOpt`, containing a sequence of the same length as assertions`, may be
```
Comment (thpani) [inline tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/stratifiedRules/aux/oracles/Oracle.scala:41]:
```suggestion
   * Optionally, a sequence option `elseAssertionsOpt`, containing a sequence of the same length as `assertions`, may be
```

**Review by thpani — COMMENTED** [review]
Looks great, I really appreciate the clarity in code + docs!

Could we add the missing unit tests here? It would make for a more self-contained PR.

File: tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/stratifiedRules/aux/oracles/IntOracle.scala (hunk 3 above)
Comment (Kukovec) [reply tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/stratifiedRules/aux/oracles/IntOracle.scala:37]: Ah, must have missed it, that was the old name. Good catch.

File: tla-bmcmt/src/test/scala/at/forsyte/apalache/tla/bmcmt/stratifiedRules/aux/oracles/TestIntOracle.scala [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+    val prop =
+      forAll(intGen) {
+        case i if i < 0 =>
+          Prop.throws(classOf[IllegalArgumentException]) {
+            IntOracle.create(initScope, i)
+          }
+        case i => IntOracle.create(initScope, i)._2.size == i
+      }
+
+    check(prop, minSuccessful(1000), sizeRange(4))
```
Comment (thpani) [inline tla-bmcmt/src/test/scala/at/forsyte/apalache/tla/bmcmt/stratifiedRules/aux/oracles/TestIntOracle.scala:47]: Can we reduce `minSuccessful` here, given that `intGen` produces at most 21 distinct integers?

**Review by thpani — APPROVED** [review]
LGTM

File: tla-bmcmt/src/test/scala/at/forsyte/apalache/tla/bmcmt/stratifiedRules/aux/oracles/TestIntOracle.scala (hunk 5 above)
Comment (Kukovec) [reply tla-bmcmt/src/test/scala/at/forsyte/apalache/tla/bmcmt/stratifiedRules/aux/oracles/TestIntOracle.scala:47]: Sure.

**Pull request merged**